### PR TITLE
Rewrote Block to mirror Vote and Timeout (Block = (BlockData, Signature))

### PR DESCRIPTION
**Pull request by davidiw**
Please ignore the first 3 commits.

- move ExecutedBlock to its own folder for readability
- replace BlockSerializer with BlockData and use that to contain the actual proposal data
- Block now contains a BlockData, the id, and an optional signature (depending on the BlockType within BlockData)

Some additional fun:
Eliminated the additional parent block used to generate a new block, we already had it as a result of the QC. This results in the elimination of a unit test.

### CODE REVIEW

**Comment by dmitri-perelman** [comment]
Sorry, I now got confused with too many Block types :).
Today we have:
* `Block` (obvious)
* `ExecutedBlock` (for ScratchPad purposes)
* `BlockSerializer` (keeping the fields to be hashed for calculating block id)

What's the idea behind `InternalBlock`?

**Comment by davidiw** [comment]
> Sorry, I now got confused with too many Block types :).
> Today we have:
> 
> * `Block` (obvious)
> * `ExecutedBlock` (for ScratchPad purposes)
> * `BlockSerializer` (keeping the fields to be hashed for calculating block id)
> 
> What's the idea behind `InternalBlock`?

This gets rid of the BlockSerializer and unifies it under internal block so that we can also use it as a container for communicating with the TCB for signing blocks.

**Comment by bmwill** [comment]
What type of "borrowing from the deserializer" were you trying? If you were trying to borrow a `&str` or a `&[u8]` then serde supports that pretty easily, if you were trying to borrow a more complex data structure then it ends up being a bit more complicated (and sometimes infeasible). Here's an [issue](https://github.com/serde-rs/serde/issues/1092) talking about references to more complicated data structures and [this](https://serde.rs/lifetimes.html) is a good reference for understanding the `'de` lifetime.

One thing that you can do to eliminate the duplicate types for `InternalBlock` is wrap the references with a [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) which Serde and the LCS impl fully supports (there's even tests for Cow's in the LCS repo).

**Comment by dmitri-perelman** [comment]
We had a great chat with @davidiw. One thing we could do is to have Block structural organization closer to what we have with Votes:
* `BlockData` doesn't keep id & signature (and kinda replaces the `BlockSerializer` by implementing CryptoHash).
* `Block` is a struct with `block_data`, `id` and `signature`.

**Comment by zekun000** [comment]
I was thinking about this `block_data`, `id`, `signature` structure too. although it doesn't play well with the current BlockType, it may still be the right way to go.

File: consensus/consensus-types/src/block.rs [hunk 1]
```diff
@@ -237,133 +156,67 @@ where
...
-            },
+            id: block_data.hash(),
+            block_data,
+            signature: None,
         }
     }
 
-    pub fn make_block(
-        parent_block: &Block<T>,
+    pub fn new_proposal(
```
Comment (dmitri-perelman) [inline consensus/consensus-types/src/block.rs]: I actually liked `make_block` more: we're generating blocks in different scenarios, e.g., retrieving the blocks from remote peers, after restart, etc., `new_proposal` sounds weird for these cases.

**Comment by dmitri-perelman** [comment]
Wow! That was pretty amazing.
@bors-libra r+

**Pull request closed without merging**